llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp

```diff
@@ -1014,6 +1015,100 @@ struct AAAMDGPUNoAGPR
 
 const char AAAMDGPUNoAGPR::ID = 0;
 
+struct AAAMDGPUInreg
```

Comment:
I don't like the name. The "inreg" attribute here is "misused" to some degree. The AA should be called "AAAMDGPUUniform" with the sub-class being "AAAMDGPUUniformArgument". The "inreg" part is really just a round-about way to mark uniform arguments.

Reply:
Well, but it is still indeed the `inreg` attribute, like the other attribute AA such as `AANoAlias`.

Reply:
I am also in favour calling it Uniform. It's an interesting choice. The attribute itself may be called ``inreg``, but it is merely an implementation of what is actually uniformity. But then, later if we want to infer ``inreg`` for some other reason Foo, I am not sure if that should be a yet another AAFoo, or it should be unified with AAUniform under a single AAInreg.

Reply:
AAUniform will determine/propagate uniformity itself at some point (see the other comment). So it actually derives uniformity, and manifests it for arguments with inreg. As you derive uniformity, other positions will require different manifestations. Still, the base property here is uniformity.

Reply:
Okay. I'm sold.